src/rio.c

```diff
@@ -518,3 +519,138 @@ size_t rioWriteBulkDouble(rio *r, double d) {
+            while(nwritten != count) {
+                retval = connWrite(r->io.connset.conns[j],p+nwritten,count-nwritten);
+                if (retval <= 0) {
+                    /* With blocking sockets, which is the sole user of this
+                     * rio target, EWOULDBLOCK is returned only because of
+                     * the SO_SNDTIMEO socket option, so we translate the error
+                     * into one more recognizable by the user. */
+                    if (retval == -1 && errno == EWOULDBLOCK) errno = ETIMEDOUT;
+                    break;
+                }
+                nwritten += retval;
+            }
+
+            if (nwritten != count) {
+                /* Mark this FD as broken. */
+                r->io.connset.state[j] = errno;
+                if (r->io.connset.state[j] == 0) r->io.connset.state[j] = EIO;
+            }
+        }
+        if (broken == r->io.connset.numconns) return 0; /* All the FDs in error. */
+        p += count;
+        len -= count;
+        r->io.connset.pos += count;
+    }
+
+    if (doflush) sdsclear(r->io.connset.buf);
+    return 1;
+}
+
+/* Returns 1 or 0 for success/failure. */
+static size_t rioConnsetRead(rio *r, void *buf, size_t len) {
+    UNUSED(r);
+    UNUSED(buf);
+    UNUSED(len);
+    return 0; /* Error, this target does not support reading. */
+}
+
+/* Returns read/write position in file. */
+static off_t rioConnsetTell(rio *r) {
+    return r->io.connset.pos;
+}
+
+/* Flushes any buffer to target device if applicable. Returns 1 on success
+ * and 0 on failures. */
+static int rioConnsetFlush(rio *r) {
+    /* Our flush is implemented by the write method, that recognizes a
+     * buffer set to NULL with a count of zero as a flush request. */
+    return rioConnsetWrite(r,NULL,0);
+}
+
+static const rio rioConnsetIO = {
+    rioConnsetRead,
+    rioConnsetWrite,
+    rioConnsetTell,
+    rioConnsetFlush,
+    NULL,           /* update_checksum */
+    0,              /* current checksum */
+    0,              /* flags */
+    0,              /* bytes read or written */
+    0,              /* read/write chunk size */
+    { { NULL, 0 } } /* union for io-specific vars */
+};
+
+
+void rioInitWithConnset(rio *r, connection **conns, int numconns) {
+    *r = rioConnsetIO;
+    r->io.connset.conns = zmalloc(sizeof(connection*) * numconns);
+    r->io.connset.state = zmalloc(sizeof(int) * numconns);
+
+    for (int i = 0;  i < numconns;  i++) {
+        connIncrRefs(conns[i]);
+        r->io.connset.conns[i] = conns[i];
+        r->io.connset.state[i] = 0;
+    }
+
+    r->io.connset.numconns = numconns;
+    r->io.connset.pos = 0;
+    r->io.connset.buf = sdsempty();
+}
+
+/* release the rio stream. */
+void rioFreeConnset(rio *r) {
+    for (int i = 0;  i < r->io.connset.numconns;  i++) {
+        connection *conn = r->io.connset.conns[i];
+        connDecrRefs(conn);
+        callHandler(conn, NULL);    // trigger close/free if necessary
```

Comment:
wrong order?
```suggestion
        callHandler(conn, NULL);    /* trigger close/free if necessary */
        connDecrRefs(conn);
```

Reply:
Firstly, we must decrease connection refs. If conn->refs does not equal zero when we get into callHandler, then the connection will not be freed.
```
    if (conn->flags & CONN_FLAG_CLOSE_SCHEDULED) {
        if (!connHasRefs(conn)) connClose(conn);
        return 0;
    }
```

Reply:
you're right.